Native windows support using clang

isuruf opened the pull request:
And CI using appveyor

tkelman commented on .appveyor.yml:
```diff
@@ -0,0 +1,27 @@
+- conda config --add channels conda-forge
+- conda install --yes clangdev posix m2-make pthreads-win32
+- copy %CONDA_INSTALL_LOCN%\Library\lib\pthreads.lib %CONDA_INSTALL_LOCN%\Library\lib\pthread.lib
+- copy %CONDA_INSTALL_LOCN%\Library\lib\pthreads.lib %CONDA_INSTALL_LOCN%\Library\lib\m.lib
```
this is pretty hacky and not something anyone will want to do locally. would be better to fix in the build system.

isuruf replied:
Fixed in https://github.com/flame/blis/pull/227/commits/a3e47eea2631c603a3f844c54d949d2025c87f4e

isuruf commented:
@fgvanzee, reordering of *mmintrin.h includes were because I get the following error otherwise,
```
In file included from kernels/penryn/1/bli_axpyv_penryn_int.c:38:
In file included from C:\Program Files\LLVM\lib\clang\6.0.0\include\pmmintrin.h:27:
In file included from C:\Program Files\LLVM\lib\clang\6.0.0\include\emmintrin.h:27:
In file included from C:\Program Files\LLVM\lib\clang\6.0.0\include\xmmintrin.h:27:
C:\Program Files\LLVM\lib\clang\6.0.0\include\mmintrin.h:81:40: fatal error: passing '__v2si' (aka 'int') to parameter of incompatible type '__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return __builtin_ia32_vec_ext_v2si((__v2si)__m, 0);
                                       ^~~~~~~~~~~
1 error generated.
Makefile:562: recipe for target 'obj/x86_64/kernels/penryn/1/bli_axpyv_penryn_int.o' failed
make: *** [obj/x86_64/kernels/penryn/1/bli_axpyv_penryn_int.o] Error 1
```

Reason why this happens is not obvious to me either. Maybe someone else with more experience can comment.

fgvanzee commented:
@isuruf In reference to your `$(LIBPTHREAD)` changes to `common.mk`: I noticed you did not convert line 338. (This line is needed because BLIS actually always needs `libpthread` regardless of whether pthread-based multithreading is enabled.) Was this omission intentional?

fgvanzee commented:
> auto (haswell) config is built fine. x86_64 fails at the ar step with,
```
Archiving lib/x86_64/libblis.a
bash: /c/Miniconda36-x64/Library/bin/llvm-ar: Argument list too long
Makefile:584: recipe for target 'lib/x86_64/libblis.a' failed
make: *** [lib/x86_64/libblis.a] Error 126
```
It would seem that the additional command line arguments from `x86_64` pushed it over the edge beyond the [max argument list limit](https://www.in-ulm.de/~mascheck/various/argmax/). I've dealt with this before in the context of libflame, and I *hated* the workaround.

Forgive me if this is a silly question, but is the Windows solution you're proposing via this pull request essentially a cross-compilation? (You compile in Linux and then it's ready to run on Windows?) What is the output of `getconf ARG_MAX` on your system?

isuruf commented:
> Forgive me if this is a silly question, but is the Windows solution you're proposing via this pull request essentially a cross-compilation? (You compile in Linux and then it's ready to run on Windows?)

No, this builds with clang running on windows and targets the MSVC ABI. (Just like clang running on linux can target glibc). It also uses `llvm-ar, llvm-as`. `clang, llvm-as, llvm-ar` all have a GNU like command line syntax, but actually creates a native windows library compatible with MSVC under the hood. This uses tools like `make, grep, sed` from msys2 project as build tools, but compilers from msys2 are not used.

> What is the output of getconf ARG_MAX on your system?

I'll run an appveyor build and let you know.

isuruf commented:
`getconf ARG_MAX` gives me `32000`

fgvanzee commented:
Thanks for explaining your build environment. That helped quite a bit.

> `getconf ARG_MAX` gives me `32000`

That's pretty small. No wonder `ar` ran out of space on the command line. Why don't set this issue aside for now since it's actually not specific to your environment/OS. I can look into implementing a `configure` option that enables an `ARG_MAX` hack along the lines of those proposed in [the link](https://www.in-ulm.de/%7Emascheck/various/argmax/) I included previously. (If it's okay with you, we don't need that code to be included in your ongoing pull request.)

fgvanzee commented:
BTW, not sure if you saw, but there appears to be an issue with the Travis CI builds:
```
configure: automatic configuration requested.
gcc-5: error: lpthread: No such file or directory
./configure: line 938: ./auto-detect.x: No such file or directory
```
Looks like the `-` is somehow getting stripped from `-lpthread`.

isuruf commented:
> Why don't set this issue aside for now since it's actually not specific to your environment/OS. I can look into implementing a configure option that enables an ARG_MAX hack along the lines of those proposed in the link I included previously. (If it's okay with you, we don't need that code to be included in your ongoing pull request.)

Works for me. In appveyor script, I changed the config to `intel64` for now. If you sign up for appveyor and add this project, all your PRs and pushes will be built on appveyor just like travis-ci, but on a windows machine.

> BTW, not sure if you saw, but there appears to be an issue with the Travis CI builds

Fixed in https://github.com/flame/blis/pull/227/commits/41f04e6a7edfea02d99b1d4a7d58d0ee14e32b65

fgvanzee commented:
BTW, regarding my `ARG_MAX` hack: one of the solutions I'm considering relies on (a) echoing the object file paths to a temporary file in whatever directory `configure` is being run, and (b) `cat` and `xargs -n`. Do you see any problems with the appveyor setup? (In particular, I wanted to be sure that `cat` and `xargs` were available, and that `xargs` accepted `-n`.)

isuruf commented:
> Do you see any problems with the appveyor setup? (In particular, I wanted to be sure that cat and xargs were available, and that xargs accepted -n.)

I don't see any problems with that. `cat` and `xargs -n` are both available

fgvanzee commented:
Okay, this looks good now. I think I'm ready to call it done.

I've added signing up for appveyor to my to-do list, so I'll look into that later. Thanks for your contributions, Isuru. Hopefully this helps out the numpy people, but even if not, I'm guessing this will be useful to others as a standalone solution.

fgvanzee commented:
Okay, one more question: What is the purpose of
```
#ifdef _MSC_VER
#define strerror_r(errno,buf,len) strerror_s(buf,len,errno)
#endif
```
in `frame/thread/bli_thread.h`?

isuruf commented:
strerror_r s not supported in windows. Found this solution at https://stackoverflow.com/questions/11718915/alternative-api-of-strerror-r-for-windows-os

fgvanzee commented:
OH, I had forgotten that we even made use of `strerror_r()` in `bli_thread.c`. Makes sense now. Thanks.

isuruf commented:
Thanks for the review and merge

fgvanzee commented:
Thanks, Nathaniel. This was another option I was looking at. I implemented a similar hack in libflame the way you suggest. However, based on some feedback from a former collaborator of ours, according to a comment he left in the `Makefile`, it didn't work on OS X, and so we had to switch to an explicit `cat`/`xargs` solution. I haven't looked into it lately--maybe it works in more recent versions of OS X--but that's why I inquired about `cat` and `xargs`.

njsmith commented on frame/base/bli_winsys.h:
```diff
@@ -0,0 +1,40 @@
+
+
+#ifdef _MSC_VER
+#include <windows.h>
```
I don't see think this should include `windows.h`. It appears to be unnecessary, and it might be harmful b/c `windows.h` dumps a *lot* of junk into your namespace, including macros with short names.

fgvanzee commented:
@njsmith Turns out I didn't even need `cat` and `xargs`. Instead, I opted for a solution that was partially in place for libflame, and seemingly endorsed by the [GNU `make` documentation](https://www.gnu.org/software/make/manual/make.html#File-Function):
```
target: prerequisites
    $(file > $@.in,$^)
    $(AR) $(ARFLAGS) $@ @$@.in
    $(RM_F) $@.in
```
@isuruf This `ARG_MAX` hack seems to work in my limited testing. I'll commit the changes so you can test them out with the `x86_64` configuration target.

fgvanzee commented:
@isuruf The `ARG_MAX` hack feature has been implemented in 2c7960c. Please let me know if it doesn't work for you.

isuruf commented:
Good news is the the arg-max-hack works. Bad news is that master is now failing after merge of dev into master.
```
kernels/sandybridge/3/bli_gemm_sandybridge_asm_d8x4.c:65:2: fatal error: unexpected token at start of statement
        begin_asm()
        ^
./frame/include\bli_x86_asm_macros.h:152:30: note: expanded from macro 'begin_asm'
#define begin_asm() BEGIN_ASM()
                             ^
```

fgvanzee commented:
Any ideas, @devinamatthews?

devinamatthews commented:
It looks like the Windows-style ASM option is being triggered, which has never been tested before. @isuruf if you don't want to be the guinea pig on this then you can disable it pretty easily  in `include/bli_x86_asm_macros.h`.

isuruf commented:
How do I do that? clang on windows supports gas syntax. https://clang.llvm.org/compatibility.html#inline-asm

devinamatthews commented:
Please change [line 52](https://github.com/flame/blis/blob/89e178ce380439dea951925e33703dc4b979e914/frame/include/bli_x86_asm_macros.h#L52) to `#if 0`.

isuruf commented:
Thanks. Fixed in https://github.com/flame/blis/pull/230
